### DISCUSSION TITLE
ActiveStorage AWS resource error on assets precompile

### DISCUSSION THREAD

**Discussion by cmaxw** [avo-hq/avo#3157] [Help]
I'm deploying with Kamal... I'm getting this error when I try to deploy

```
bundle exec rails assets:precompile
#13 4.658 Error retrieving instance profile credentials: Failed to open TCP connection to 169.254.169.254:80 (Connection refused - connect(2) for "169.254.169.254" port 80)
#13 4.684 bin/rails aborted!
#13 4.697 ArgumentError: missing required option :name (ArgumentError)
#13 4.697
#13 4.697       when nil then raise ArgumentError, "missing required option :name"
#13 4.697                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#13 4.697 /rails/app/models/meetup.rb:19:in `<class:Meetup>'
#13 4.697 /rails/app/models/meetup.rb:18:in `<main>'
#13 4.697 /rails/app/avo/resources/meetup_replay.rb:4:in `<class:MeetupReplay>'
#13 4.697 /rails/app/avo/resources/meetup_replay.rb:1:in `<main>'
#13 4.697 /rails/config/environment.rb:7:in `<main>'
#13 4.697 Tasks: TOP => assets:precompile => environment
#13 4.697 (See full trace by running task with --trace)
#13 ERROR: process "/bin/sh -c SECRET_KEY_BASE_DUMMY=1 bundle exec rails assets:precompile" did not complete successfully: exit code: 1
------
 > [8/8] RUN SECRET_KEY_BASE_DUMMY=1 bundle exec rails assets:precompile:
4.697
4.697       when nil then raise ArgumentError, "missing required option :name"
4.697                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
4.697 /rails/app/models/meetup.rb:19:in `<class:Meetup>'
4.697 /rails/app/models/meetup.rb:18:in `<main>'
4.697 /rails/app/avo/resources/meetup_replay.rb:4:in `<class:MeetupReplay>'
4.697 /rails/app/avo/resources/meetup_replay.rb:1:in `<main>'
4.697 /rails/config/environment.rb:7:in `<main>'
4.697 Tasks: TOP => assets:precompile => environment
4.697 (See full trace by running task with --trace)
------
Dockerfile:49
--------------------
  47 |
  48 |     # Precompiling assets for production without requiring secret RAILS_MASTER_KEY
  49 | >>> RUN SECRET_KEY_BASE_DUMMY=1 bundle exec rails assets:precompile
  50 |
  51 |     # Entrypoint prepares the database.
--------------------
```

The issue tracks back to this resource:

```
class Avo::Resources::MeetupReplay < Avo::BaseResource
  # self.includes = []
  # self.attachments = []
  self.model_class = ::Meetup
  self.index_query = -> { query.past }
  self.default_view_type = :grid
  # self.search = {
  #   query: -> { query.ransack(id_eq: params[:q], m: "or").result(distinct: false) }
  # }

  def fields
    field :id, as: :idlo
    field :title, as: :text
    field :slug, as: :text
    field :description, as: :textarea
    field :starts_at, as: :date_time
    field :video_embed, as: :textarea
    field :participant_url, as: :text
    field :location, as: :text
    field :thumbnail, as: :file
    field :authors, as: :has_many, through: :meetup_authors
    field :membership_level_meetups, as: :has_many
    field :membership_levels, as: :has_many, through: :membership_level_meetups
    field :transcript, as: :trix
    field :notes, as: :trix
  end

  self.grid_view = {
    card: -> do
      {
        cover_url:
          if record.thumbnail.attached?
            main_app.url_for(record.thumbnail.url)
          end,
        title: record.title,
        body: record.description&.truncate(60, separator: /\s/)
      }
    end
  }
end
```

The error points to the `has_one_attached :thumbnail`

It works on the meetup resource, but not on the MeetupReplay resource. The only difference between the two files is the `model_class` and the `index_query` scope.

It's also working in development, which uses `:disk` for storage instead of an AWS compatible storage.

**Comment by adrianthedev** [comment]
Hey @cmaxw please post the model source code as well.

I see that the stack trace points to the model, not the resource.
I would be inclined to say it's not an Avo problem.

**Reply by adrianthedev** [reply]
Also, try to set the `model_class` to a string instead of the class name.

```ruby
# instead of 
self.model_class = ::Meetup

# do this
self.model_class = "Meetup"
```

The initialization of that model might be the issue.

Please let us know if this fixes the issue.

**Reply by cmaxw** [reply]
The line it points to in the model is `has_one_attached :thumbnail`

The resource is identical to the meetup resource except the name and the scope. This error occurs only when I have this resource in the mix. I'll try assigning it as a string to see if that solves it.

**Reply by cmaxw** [reply]
That does not fix it. It gives the same error.

Here's the Meetup model.

```ruby
# frozen_string_literal: true

# == Schema Information
#
# Table name: meetups
#
#  id              :bigint           not null, primary key
#  description     :text
#  location        :string
#  participant_url :string
#  slug            :string
#  starts_at       :datetime
#  title           :string
#  video_embed     :text
#  created_at      :datetime         not null
#  updated_at      :datetime         not null
#
class Meetup < ApplicationRecord
  has_one_attached :thumbnail
  has_rich_text :transcript
  has_rich_text :notes

  has_many :meetup_authors
  has_many :authors, through: :meetup_authors
  has_many :membership_level_meetups
  has_many :membership_levels, through: :membership_level_meetups

  scope :upcoming, -> { where('starts_at >= ?', 1.hour.ago).order(starts_at: :asc) }
  scope :past, -> { where('starts_at < ?', 1.hour.ago).order(starts_at: :desc) }

  extend FriendlyId
  friendly_id :title, use: :slugged
end
```

**Reply by adrianthedev** [reply]
Hmmm...

I think I found it.

It seems to be related to the AWS SDK.
It's the [only place](https://github.com/search?q=repo%3Aaws%2Faws-sdk-ruby%20%22missing%20required%20option%20%3Aname%22&type=code) where I could find that error.

Please check your AWS credentials. If you fix those, then it should pass.

**Reply by cmaxw** [reply]
Everything works fine when I take that resource out. I can even upload new images and put them into cloud storage.

**Reply by cmaxw** [reply]
So, the credentials work.

**Reply by adrianthedev** [reply]
Hmmm. ok.
My hunch is that something is triggering the model to connect to AWS and probably the credentials are not loaded/available in that moment.

So you're saying that other resources that have active storage files attached and don'te have the `model_class` option work just fine.

Let me dig into it a bit more.
Trying to see if `model_class` makes things weird.

**Reply by Paul-Bob** [reply]
> Everything works fine when I take that resource out. I can even upload new images and put them into cloud storage.

Maybe comment element by element to identify which field or configuration is causing the error (I guess `field :thumbnail, as: :file`).

I just noticed that your `:id` field is declared as `:idlo`, which is a non-existent type. While this might not be the cause of the issue, you should change it to `field :id, as: :id`.

**Reply by Paul-Bob** [reply]
Is there any option that configures a bucket name? The error complains about missing `:name` (`"missing required option :name"`)

**Comment by cmaxw** [comment]
Correct. I have a Meetup resource that is the same except the `index_query` and `model_class`

**Reply by adrianthedev** [reply]
Got it, Charles!

We're going to be out  on vacation next week but we'll try to debug this when we get some spare time.

**Reply by adrianthedev** [reply]
Actually, please add this bit of code somewhere in an initializer.
Could be `avo.rb` at the bottom.

```ruby
module Avo
  module Resources
    class Base
      def self.model_class=(value)
        @model_class = case value
        when Class
          value
        when String, Symbol
          value.to_s
        else
          raise ArgumentError.new "Failed to find a proper model class for #{self}"
        end
      end
    end
  end
end
```

I'm trying to see if this will make that step of the way pass. Avo might fail somewhere else, but at least we validated that this was causing your error.

**Comment by cmaxw** [comment]
It deploys now... but I'm getting this error when I load the page:

`unable to find policy `StringPolicy` for `"Meetup"` (Avo::NoPolicyError)`

**Comment by cmaxw** [comment]
I get the same when I have the string "::Meetup"

**Comment by cmaxw** [comment]
If I switch it back to a class like `Meetup` or `::Meetup` it gives the same error there, even with the initializer changed up.

**Comment by cmaxw** [comment]
I finally gave up and created a new model. I could also have done it with scopes if I upgraded to Advanced... but Replays are different enough from Meetups that I'm fine with what I've got.

**Reply by adrianthedev** [reply]
Hmmm. This isn't what we intended. Your domain model should not have to suffer because of Avo.

I think I found the "bug". We are trying to constantize the `model_class` [when we set it](https://github.com/avo-hq/avo/blob/main/lib/avo/concerns/model_class_constantized.rb#L13) and that's triggering the error.
We will refactor it to constantize it when it's requested. I'm opening an issue for that.

Thanks for the troubleshooting session.

**Comment by adrianthedev** [comment]
Fixed by https://github.com/avo-hq/avo/pull/3207
